### ISSUE TITLE
[autoscaler] Where to add container registry authentication?

### ISSUE THREAD

**Issue by ca-scribner** [ray-project/ray#7610]
I'm trying to set up autoscaler to use a docker container from google container registry, which requires authentication (credentials can be obtained on the host machine by running `docker-credential-gcr configure-docker` before pulling the container).  Where should this code be put in the autoscaler yaml?  I thought it would be in initialization_commands like this:

`
initialization_commands: [
  "docker-credential-gcr configure-docker",
]
`

But I'm getting authentication rejections which make me think that's not working (not sure how to dig deeper and fully understand the error though).  Does this seem like the correct workflow or that I'm missing something obvious?  Thanks!

**Comment by anabranch** [comment]
Hey andrew, thanks for this question. are you running the autoscaler or running kubernetes or how are you managing the cluster? Would you be open to having a more detailed chat about this so we can better understand your use case?

**Comment by richardliaw** [comment]
(Also sorry for the late reply; the initialization looks right. If you can post a stack trace, that'd be helpful for us to debug.)

**Comment by ca-scribner** [comment]
No problem on slow reply - glad to hear from you both.

@anabranch a chat might make more sense.  Thanks to some corvid-related isolation, I dove pretty deep into this.  My gut feel is that I'm either ignorant and misusing the autoscaler or I'm doing something different from what was expected and finding some new edge cases.  Either way, happy to make some PRs with what I've found or help someone else to fix/extend. 

Richard added me to the slack - want to dig into things there when you have a minute?

**Comment by IsaacWolverton** [comment]
@ca-scribner I am running into the same issue, did you ever find a solution?

**Comment by ca-scribner** [comment]
Kinda.  

To get things running how I wanted I had to make some [modifications](https://github.com/ca-scribner/ray/tree/gcp_updates).  They are a few weeks behind the main code but still should roughly show what needed to change.  I also spoke with @anabranch about it so maybe some of this has been addressed in the main code now?

If I was doing this again from scratch knowing what I know, I think I'd avoid using containers and instead leverage GCP VM images + git pulls.  Using images felt more mature than containers in Ray, and at the same time your scale up time will be much faster (even when your container is in GCR, it still takes some time to pull if it is decent size).  I'd script something that lets me easily push new changes to the VM image from my machine so it was about as easy as building new containers, then stick to that.

That being said, this is all an opinion from March 2020 - since the project is active, anyone reading this in future should do their homework.
